crates/oxc_minifier/src/peephole/remove_unused_expression.rs

```diff
@@ -562,50 +563,71 @@ impl<'a> PeepholeOptimizations {
             return false;
         }
 
+        !e.may_have_side_effects(ctx)
+    }
+
+    pub fn fold_iife_expression(e: &mut Expression<'a>, ctx: &mut Ctx<'a, '_>) {
+        let Expression::CallExpression(call_expr) = e else { return };
+
         if call_expr.arguments.is_empty() {
             let is_empty_iife = match &call_expr.callee {
                 Expression::FunctionExpression(f) => {
-                    f.params.is_empty() && f.body.as_ref().is_some_and(|body| body.is_empty())
+                    f.params.is_empty()
+                        && f.body.as_ref().is_some_and(|body| body.is_empty())
+                        && (!f.r#async && !f.generator
+                            || ctx.ancestors().nth(1).is_some_and(Ancestor::is_parent_of_statement))
+                }
+                Expression::ArrowFunctionExpression(f) => {
+                    f.params.is_empty()
+                        && f.body.is_empty()
+                        && (!f.r#async
+                            || ctx.ancestors().nth(1).is_some_and(Ancestor::is_parent_of_statement))
                 }
-                Expression::ArrowFunctionExpression(f) => f.params.is_empty() && f.body.is_empty(),
                 _ => false,
             };
             if is_empty_iife {
-                return true;
+                *e = ctx.value_to_expr(call_expr.span, ConstantValue::Undefined);
+                ctx.state.changed = true;
+                // Replace "(() => {})()" with "undefined"
+                // Replace "(function () => { return })()" with "undefined"
+                return;
             }
 
             if let Expression::ArrowFunctionExpression(f) = &mut call_expr.callee
                 && !f.r#async
-                && f.params.parameters_count() == 0
+                && !f.params.has_parameter()
                 && f.body.statements.len() == 1
             {
                 if f.expression {
                     // Replace "(() => foo())()" with "foo()"
                     let expr = f.get_expression_mut().unwrap();
                     *e = expr.take_in(ctx.ast);
-                    return Self::remove_unused_expression(e, ctx);
+                    ctx.state.changed = true;
+                    return;
                 }
                 match &mut f.body.statements[0] {
                     Statement::ExpressionStatement(expr_stmt) => {
-                        // Replace "(() => { foo() })" with "foo()"
-                        *e = expr_stmt.expression.take_in(ctx.ast);
-                        return Self::remove_unused_expression(e, ctx);
+                        // Replace "(() => { foo() })" with "(foo(), undefined)"
```

Comment:
```suggestion
                        // Replace "(() => { foo() })()" with "(foo(), undefined)"
```
(was a typo before this PR)